src/entry.rs

```diff
@@ -12,61 +12,50 @@ use crate::{
 };
 
 #[derive(Debug, Clone)]
-pub enum EntryKind {
-    File,
-    Dir,
-    NonExist,
-    Unknown,
+pub struct EntryStat {
+    /// `None` for non-existing file
+    file_type: Option<FileType>,
+
+    /// `None` for existing file but without system time.
+    modified: Option<SystemTime>,
 }
 
-impl EntryKind {
-    pub fn is_file(&self) -> bool {
-        matches!(self, EntryKind::File)
+impl EntryStat {
+    fn new(file_type: Option<FileType>, modified: Option<SystemTime>) -> Self {
+        Self {
+            file_type,
+            modified,
+        }
     }
 
-    pub fn is_dir(&self) -> bool {
-        matches!(self, EntryKind::Dir)
+    /// Returns `None` for existing file but without system time.
+    pub fn modified(&self) -> Option<&SystemTime> {
+        self.modified.as_ref()
     }
 
-    pub fn exist(&self) -> bool {
-        matches!(self, EntryKind::NonExist)
+    /// Returns `None` for non-existing file
+    pub fn file_type(&self) -> Option<&FileType> {
+        self.file_type.as_ref()
     }
-}
 
-#[derive(Debug, Clone)]
-pub struct EntryStat {
-    pub kind: EntryKind,
-    pub mtime: Option<SystemTime>,
-}
-
-impl EntryStat {
-    pub fn stat(path: &Path) -> io::Result<Self> {
-        let stat = if let Ok(meta) = std::fs::metadata(path) {
-            let kind = if meta.is_file() {
-                EntryKind::File
-            } else if meta.is_dir() {
-                EntryKind::Dir
-            } else {
-                EntryKind::Unknown
-            };
-            let mtime = Some(meta.modified()?);
-            EntryStat { kind, mtime }
+    fn stat(path: &Path) -> Self {
+        if let Ok(meta) = path.metadata() {
+            // This field might not be available on all platforms,
+            // and will return an Err on platforms where it is not available.
+            let modified = meta.modified().ok();
+            Self::new(Some(meta.file_type()), modified)
         } else {
-            EntryStat {
-                kind: EntryKind::NonExist,
-                mtime: None,
-            }
-        };
-        Ok(stat)
+            Self::new(None, None)
+        }
     }
 }
 
 #[derive(Debug)]
 pub struct Entry {
-    pub parent: Option<Arc<Entry>>,
-    pub path: PathBuf,
-    pub pkg_info: Option<Arc<PkgInfo>>,
-    pub stat: RwLock<Option<EntryStat>>,
+    parent: Option<Arc<Entry>>,
+    path: Box<Path>,
+    pkg_info: Option<Arc<PkgInfo>>,
+    stat: RwLock<Option<Arc<EntryStat>>>,
```

Comment:
why wrapped by `Arc`?

Reply:
No longer needed, EntryStat is `Copy` now.